homeassistant/components/alexa/auth.py

```diff
@@ -0,0 +1,148 @@
+"""Support for Alexa skill auth."""
+
+import asyncio
+import json
+import logging
+from datetime import timedelta
+import aiohttp
+import async_timeout
+
+from homeassistant.core import callback
+from homeassistant.helpers import aiohttp_client
+from homeassistant.util import dt
+from .const import DEFAULT_TIMEOUT
+
+_LOGGER = logging.getLogger(__name__)
+
+LWA_TOKEN_URI = "https://api.amazon.com/auth/o2/token"
+LWA_HEADERS = {
+    "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8"
+}
+
+PREEMPTIVE_REFRESH_TTL_IN_SECONDS = 300
+STORAGE_KEY = 'alexa_auth'
+STORAGE_VERSION = 1
+STORAGE_EXPIRE_TIME = "expire_time"
+STORAGE_ACCESS_TOKEN = "access_token"
+STORAGE_REFRESH_TOKEN = "refresh_token"
+
+
+class Auth:
+    """Handle authentication to send events to Alexa."""
+
+    def __init__(self, hass, client_id, client_secret):
+        """Initialize the Auth class."""
+        self.hass = hass
+
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+        self._prefs = None
+        self._store = hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+    async def async_do_auth(self, accept_grant_code):
+        """Do authentication with an AcceptGrant code."""
+        # access token not retrieved yet for the first time, so this should
+        # be an access token request
+
+        lwa_params = {
+            "grant_type": "authorization_code",
+            "code": accept_grant_code,
+            "client_id": self.client_id,
+            "client_secret": self.client_secret
+        }
+        _LOGGER.debug("Calling LWA to get the access token (first time), "
+                      "with: %s", json.dumps(lwa_params))
+
+        return await self._async_request_new_token(lwa_params)
+
+    async def async_get_access_token(self):
+        """Perform access token or token refresh request."""
+        if self._prefs is None:
```

Comment:
We should wrap this whole method in an async lock to avoid requesting multiple tokens at once.